### PR TITLE
Replace std::regex with fnmatch()/PathMatchSpec as a workaround to std::regex stack overflow known bug

### PR DESCRIPTION
**Pull request by rkamd**
Resolves : [SWDEV-427140](https://ontrack-internal.amd.com/browse/SWDEV-427140),

An underlying bug in std::regex ( libstdc++) is causing stack overflow and crashes the application. 

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86164#c8 

This PR replaces the use of std::regex_search() with fnmatch().

### CODE REVIEW

File: Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
 #include <Tensile/MasterSolutionLibrary.hpp>
 #include <Tensile/PlaceholderLibrary.hpp>
-#include <regex>
+#include <fnmatch.h>
```
Comment (daineAMD) [inline Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:34]: Might need windows-specific code?

Comment (rkamd) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:34]: good catch, I will add support for windows

Comment (TorreZuk) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:34]: Yes for windows sake as well did you consider what I suggested in the meeting?   Can the requirement be represented in simpler terms.  If you can put it in words first, it may be just two calls: strstr and maybe a strncmp suffice for what you need.

File: Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp [hunk 2]
```diff
@@ -68,7 +74,11 @@ namespace Tensile
                     for(auto condition : ctx->preloaded)
                     {
                         std::string pattern = RegexPattern(condition);
-                        if(std::regex_search(lib.filePrefix, std::regex(pattern)))
+#ifdef WIN32
+                        if(PathMatchSpec(lib.filePrefix.c_str(), pattern.c_str()))
+#else
```
Comment (TorreZuk) [inline Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:79]: Double check if tensile is using specific char set form to avoid runtime errors if build type changes, i.e. PathMatchSpecA may be preferred.   Also appears shlwapi usage varies a lot across rocm, delay load dll to static and other forms of use may cause deployment runtime errors.  See if you can find out from other libraries before adding this new dependency.

File: Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp (hunk 1 above)
Comment (rkamd) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:39]: Addressed windows version for pattern matching. @TorreZuk , Sorry I missed this comment. For the current use case, I think it could be done with just strstr. 
- strstr(lib_filename , ArchName) for arch specific 
- strstr(lib_filename, "") for loading all dat files.

My thought process to replace it with an equivalent pattern matching is, if lazy loading is expanded to load datatype specific or something like that, it would be easier to do it with pattern matching. 

I could use strstr if that is preferred.

Comment (TorreZuk) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp:39]: if !strlen(pattern) you load them all so that is simple.  Yes if you expect to need a full glob later then strstr may get too tricky.  The way you link did it bring in a .dll or a static library for sh?   We then static link this into rocblas.dll so confirm rocblas dependencies on windows and if it passes tests that is good enough for me.  Just saying it appears there is variation in consuming this library on windows so something to check out.

File: Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp (hunk 2 above)
Comment (rkamd) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp]: shlwapi is used similar to how it is linked in this PR across other libraries. onnxrt uses delayload, but I do not think it might be needed in this case. I have compiled rocblas on windows with this change and it compiles fine and the pattern matching works as expected.

File: Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp [hunk 3]
```diff
@@ -31,7 +31,13 @@
 
 #include <Tensile/MasterSolutionLibrary.hpp>
 #include <Tensile/PlaceholderLibrary.hpp>
-#include <regex>
+//Replace std::regex, as it crashes when matching long lines(GCC Bug #86164).
+#ifdef WIN32
+#include "shlwapi.h"
+#include "windows.h"
```
[the remark this answers is not in the document]
Comment (rkamd) [reply Tensile/Source/lib/include/Tensile/Serialization/PlaceholderLibrary.hpp]: @AlexBrownAMD ,Removed windows.h , as it was not required

**Review by AlexBrownAMD — APPROVED** [review]

**Review by TorreZuk — APPROVED** [review]
If the windows rocblas build can load tensile kernels then LGTM
